src/api/v4/layer/layer.js

```diff
@@ -53,22 +53,35 @@ Layer.prototype = Object.create(Base.prototype);
  *
  * @param {carto.style.CartoCSS} New style
  * @fires carto.layer.Layer.styleChanged
- * @return {carto.layer.Layer} this
- *
+ * @return {Promise} A promise that will be fulfilled when the style is applied to the layer or rejected with a
+ * {@link CartoError} if something goes bad
  * @api
  */
 Layer.prototype.setStyle = function (style, opts) {
+  var self = this;
   var prevStyle = this._style;
   _checkStyle(style);
   opts = opts || {};
-  this._style = style;
-  if (this._internalModel) {
-    this._internalModel.set('cartocss', style.toCartoCSS());
+  if (prevStyle === style) {
+    return;
   }
-  if (prevStyle !== style) {
+  if (this._internalModel) {
+    this._internalModel.set('cartocss', style.toCartoCSS(), { silent: true });
+    return this._engine.reload()
+      .then(function () {
+        self._style = style;
+        self.trigger('styleChanged', this);
+      })
+      .catch(function (err) {
+        var error = new CartoError(err);
+        self.trigger('error', new CartoError(error));
+        return Promise.reject(error);
```

Comment:
Why do we `reject` here but don't `resolve` in the `then` function?

Reply:
Spoiler: worst explanation ever:

Since promises are designed to be chained, everything you return inside a `.then` gets wrapped into a promise.  

If the function dont return anything the function `"undefined"` is returned wrapped into a promise. Since [engine.reload is resolved with no parameters](https://github.com/CartoDB/cartodb.js/blob/v4/src/engine.js#L139), we are adding a "middleware" which throws events to this promise there is no side effects.